<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tools to interact with NCBI's Pathogen Detection project • pdtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Tools to interact with NCBI's Pathogen Detection project">
<meta property="og:description" content="A collection of functions to help work with data from the ncbi pathogen detection project/database.
    Has functions for downloading the most recent metadata, cleaning metadata, downloading genome assemblies,
    and setting up and working with pangenomes.  ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">pdtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="pdtools">pdtools<a class="anchor" aria-label="anchor" href="#pdtools"></a>
</h1></div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><em>Requires R &gt;= 4.1</em> <em>Requires biostrings, need to install from biocmanager</em></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">'Biostrings'</span><span class="op">)</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">'jtrachsel/pdtools'</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h2>
<p>A collection of functions for working with data from the <a href="https://www.ncbi.nlm.nih.gov/pathogens/" class="external-link">NCBI Pathogen Detection project</a></p>
<p><a href="https://jtrachsel.github.io/pdtools/index.html" class="external-link uri">https://jtrachsel.github.io/pdtools/index.html</a></p>
</div>
<div class="section level2">
<h2 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h2>
<div class="section level4">
<h4 id="list-available-organisms">List available organisms<a class="anchor" aria-label="anchor" href="#list-available-organisms"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pdtools</span><span class="op">)</span>

<span class="fu"><a href="reference/list_organisms.html">list_organisms</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="download-the-most-recent-metadata-for-an-organism">Download the most recent metadata for an organism:<a class="anchor" aria-label="anchor" href="#download-the-most-recent-metadata-for-an-organism"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">'mkdir data'</span><span class="op">)</span>
<span class="fu"><a href="reference/download_most_recent_complete.html">download_most_recent_complete</a></span><span class="op">(</span><span class="st">'Campylobacter'</span>, folder_prefix <span class="op">=</span> <span class="st">'./data/'</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="join-downloads-metadata-and-snp-clusters">Join downloads: metadata and SNP clusters<a class="anchor" aria-label="anchor" href="#join-downloads-metadata-and-snp-clusters"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># The names of these files will change based on the most recent complete data</span>
<span class="co"># you will have to adjust these</span>
<span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">'./data/PDG000000003.1540.amr.metadata.tsv'</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">'./data/PDG000000003.1540.cluster_list.tsv'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="extract-agricultural-host-species-from-several-metadata-fields">Extract agricultural host species from several metadata fields:<a class="anchor" aria-label="anchor" href="#extract-agricultural-host-species-from-several-metadata-fields"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create a two column tibble containing consensus host for each isolate</span>
<span class="va">host_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/extract_consensus_ag_species.html">extract_consensus_ag_species</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span>

<span class="co"># join back to metadata</span>
<span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">host_info</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="extract-earliest-year-from-3-date-columns">Extract earliest year from 3 date columns<a class="anchor" aria-label="anchor" href="#extract-earliest-year-from-3-date-columns"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">earliest_year</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/extract_earliest_year.html">extract_earliest_year</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">earliest_year</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="generate-ftp-download-paths-for-a-collection-of-isolates">Generate ftp download paths for a collection of isolates<a class="anchor" aria-label="anchor" href="#generate-ftp-download-paths-for-a-collection-of-isolates"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># download most recent assembly summary</span>
<span class="fu">curl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/curl/man/curl_download.html" class="external-link">curl_download</a></span><span class="op">(</span><span class="st">'https://ftp.ncbi.nlm.nih.gov/genomes/genbank/bacteria/assembly_summary.txt'</span>,
                    destfile <span class="op">=</span> <span class="st">'./data/assembly_summary.txt'</span><span class="op">)</span>

<span class="co"># only download isolates from swine after 2015</span>
<span class="va">meta_filt</span> <span class="op">&lt;-</span>
  <span class="va">meta</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Year</span> <span class="op">&gt;</span> <span class="fl">2015</span> <span class="op">&amp;</span> <span class="va">ag_match</span> <span class="op">==</span> <span class="st">'Swine'</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">write_tsv</span><span class="op">(</span><span class="st">'./data/swine_2015_meta.tsv'</span><span class="op">)</span>

<span class="va">ftp_paths</span> <span class="op">&lt;-</span> 
  <span class="va">meta_filt</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/make_fna_urls.html">make_fna_urls</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">asm_acc</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">write_lines</span><span class="op">(</span><span class="st">'./data/ftp_download_paths.txt'</span><span class="op">)</span>

<span class="co"># download from the command line:</span>
<span class="co"># cd data</span>
<span class="co"># cat ftp_download_paths.txt | parallel -j 2 'wget {}'</span>

<span class="co"># or within R:</span>
<span class="co"># set names to the desired path </span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ftp_paths</span><span class="op">)</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'./data/'</span>,<span class="va">meta_filt</span><span class="op">$</span><span class="va">asm_acc</span>, <span class="st">'.fna'</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Map</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">u</span>, <span class="va">d</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">u</span>, <span class="va">d</span>, mode<span class="op">=</span><span class="st">"wb"</span><span class="op">)</span>, <span class="va">ftp_paths</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ftp_paths</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="generate-an-input-file-for-caclulating-a-pangenome-with-ppanggolin">Generate an input file for caclulating a pangenome with <a href="https://github.com/labgem/PPanGGOLiN" class="external-link">ppanggolin</a>
<a class="anchor" aria-label="anchor" href="#generate-an-input-file-for-caclulating-a-pangenome-with-ppanggolin"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># if you have some reference genomes that are complete (circularized) you can </span>
<span class="co"># feed their paths into the 'complete_genome_paths parameter and the function</span>
<span class="co"># will correctly specify cirular contigs for ppanggolin.  </span>

<span class="va">fna_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">'./data/'</span>, <span class="st">'.fna'</span>, full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>

<span class="fu"><a href="reference/build_ppanggolin_file_fastas.html">build_ppanggolin_file_fastas</a></span><span class="op">(</span>incomplete_genome_paths <span class="op">=</span> <span class="va">fna_files</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">write_tsv</span><span class="op">(</span><span class="st">'ppanggolin_file.tsv'</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="select-a-representative-set-of-isolates-from-a-pangenome">Select a representative set of isolates from a pangenome<a class="anchor" aria-label="anchor" href="#select-a-representative-set-of-isolates-from-a-pangenome"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Read in presence/absence matrix and format correctly:</span>
<span class="va">pan_PA</span> <span class="op">&lt;-</span>
  <span class="fu">read_tsv</span><span class="op">(</span><span class="st">'./pan/gene_presence_absence.Rtab'</span><span class="op">)</span>  <span class="op">%&gt;%</span> 
  <span class="fu">column_to_rownames</span><span class="op">(</span>var <span class="op">=</span> <span class="st">'Gene'</span><span class="op">)</span>  <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> 

<span class="co"># this will return a a small set of genomes that contain at least the proportion of</span>
<span class="co"># genes you specify</span>

<span class="co"># this set of genomes will contain 99% of all the genes detected in the pangenome</span>
<span class="fu"><a href="reference/get_pangenome_representatives.html">get_pangenome_representatives</a></span><span class="op">(</span>pan_mat <span class="op">=</span> <span class="va">pan_PA</span>, SEED <span class="op">=</span> <span class="fl">2</span>, desired_coverage <span class="op">=</span> <span class="fl">.99</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="todo">TODO<a class="anchor" aria-label="anchor" href="#todo"></a>
</h2>
<ul>
<li>update_collection() function?
<ul>
<li>Should take and old metadata file and a new metadata file as inputs.<br>
</li>
<li>return a vector of new genomes that were not present in the old list.</li>
<li>also return genomes with newer assembly accession version<br>
</li>
</ul>
</li>
<li>extract_consensus_ag_species currently assumes isolates are from humans if the epi_type is clinical and no other information is available. This is probably wrong in some cases…</li>
<li>Reference README? <a href="https://ftp.ncbi.nlm.nih.gov/pathogen/ReadMe.txt" class="external-link uri">https://ftp.ncbi.nlm.nih.gov/pathogen/ReadMe.txt</a>
</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3) + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing pdtools</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Julian Trachsel <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/https://orcid.org/0000-0003-2357-7737" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Julian Trachsel.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
