<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="A package to work with data from the
    ncbi pathogen detection project/database.  Has functions for
    downloading the most recent metadata, cleaning metadata, downloading
    genome assemblies, and setting up and working with pangenomes.">
<title>Tools to interact with NCBI's Pathogen Detection project â€¢ pdtools</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Tools to interact with NCBI's Pathogen Detection project">
<meta property="og:description" content="A package to work with data from the
    ncbi pathogen detection project/database.  Has functions for
    downloading the most recent metadata, cleaning metadata, downloading
    genome assemblies, and setting up and working with pangenomes.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">pdtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Jtrachsel/pdtools/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="pdtools">pdtools<a class="anchor" aria-label="anchor" href="#pdtools"></a>
</h1></div>
<!-- badges: start -->
<!-- badges: end --><div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">'jtrachsel/pdtools'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h2>
<p>A collection of functions for working with data from the <a href="https://www.ncbi.nlm.nih.gov/pathogens/" class="external-link">NCBI Pathogen Detection project</a></p>
</div>
<div class="section level2">
<h2 id="quickstart">Quickstart<a class="anchor" aria-label="anchor" href="#quickstart"></a>
</h2>
<div class="section level4">
<h4 id="list-available-organisms">List available organisms<a class="anchor" aria-label="anchor" href="#list-available-organisms"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jtrachsel.github.io/pdtools/">pdtools</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/list_organisms.html">list_organisms</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="download-the-most-recent-metadata-for-an-organism">Download the most recent metadata for an organism:<a class="anchor" aria-label="anchor" href="#download-the-most-recent-metadata-for-an-organism"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">'mkdir data'</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/download_most_recent_complete.html">download_most_recent_complete</a></span><span class="op">(</span><span class="st">'Campylobacter'</span>, folder_prefix <span class="op">=</span> <span class="st">'./data/'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="join-downloads-metadata-and-snp-clusters">Join downloads: metadata and SNP clusters<a class="anchor" aria-label="anchor" href="#join-downloads-metadata-and-snp-clusters"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The names of these files will change based on the most recent complete data</span></span>
<span><span class="co"># you will have to adjust these</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">'./data/PDG000000003.1540.amr.metadata.tsv'</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">'./data/PDG000000003.1540.cluster_list.tsv'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="extract-agricultural-host-species-from-several-metadata-fields">Extract agricultural host species from several metadata fields:<a class="anchor" aria-label="anchor" href="#extract-agricultural-host-species-from-several-metadata-fields"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a two column tibble containing consensus host for each isolate</span></span>
<span><span class="va">host_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/extract_consensus_ag_species.html">extract_consensus_ag_species</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># join back to metadata</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> <span class="fu">left_join</span><span class="op">(</span><span class="va">host_info</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="extract-earliest-year-from-3-date-columns">Extract earliest year from 3 date columns<a class="anchor" aria-label="anchor" href="#extract-earliest-year-from-3-date-columns"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">earliest_year</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="reference/extract_earliest_year.html">extract_earliest_year</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> <span class="fu">left_join</span><span class="op">(</span><span class="va">earliest_year</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="generate-ftp-download-paths-for-a-selection-of-isolates">Generate ftp download paths for a selection of isolates<a class="anchor" aria-label="anchor" href="#generate-ftp-download-paths-for-a-selection-of-isolates"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download most recent assembly summary</span></span>
<span><span class="fu"><a href="reference/download_gbk_assembly_summary.html">download_gbk_assembly_summary</a></span><span class="op">(</span><span class="st">'./data/assembly_summary.txt'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select isolates associated with swine from after 2015</span></span>
<span><span class="va">meta_filt</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">meta</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Year</span> <span class="op">&gt;</span> <span class="fl">2015</span> <span class="op">&amp;</span> <span class="va">ag_match</span> <span class="op">==</span> <span class="st">'Swine'</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">write_tsv</span><span class="op">(</span><span class="st">'./data/swine_2015_meta.tsv'</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># make a 'download data' tibble to organize and track downloads</span></span>
<span><span class="co"># downloads both fna and gff files for each genome in the metadata</span></span>
<span><span class="va">download_data</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">meta_filt</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">asm_acc</span>, <span class="va">ftp_path</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="reference/make_ftp_paths.html">make_ftp_paths</a></span><span class="op">(</span>assembly_summary_path <span class="op">=</span> <span class="st">'./data/assembly_summary.txt'</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="reference/make_download_urls.html">make_download_urls</a></span><span class="op">(</span><span class="st">'fna'</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="reference/make_download_urls.html">make_download_urls</a></span><span class="op">(</span><span class="st">'gff'</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="reference/make_dest_paths.html">make_dest_paths</a></span><span class="op">(</span>type<span class="op">=</span><span class="st">'fna'</span>, dest_dir <span class="op">=</span> <span class="st">'./data/'</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="reference/make_dest_paths.html">make_dest_paths</a></span><span class="op">(</span>type<span class="op">=</span><span class="st">'gff'</span>, dest_dir <span class="op">=</span> <span class="st">'./data/'</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="reference/download_genomes.html">download_genomes</a></span><span class="op">(</span><span class="st">'fna'</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="reference/download_genomes.html">download_genomes</a></span><span class="op">(</span><span class="st">'gff'</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">write_tsv</span><span class="op">(</span><span class="st">'./data/download_data.tsv'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="generate-an-input-file-for-caclulating-a-pangenome-with-ppanggolin">Generate an input file for caclulating a pangenome with <a href="https://github.com/labgem/PPanGGOLiN" class="external-link">ppanggolin</a>
<a class="anchor" aria-label="anchor" href="#generate-an-input-file-for-caclulating-a-pangenome-with-ppanggolin"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># if you have some reference genomes that are complete (circularized) you can </span></span>
<span><span class="co"># feed their paths into the 'complete_genome_paths parameter and the function</span></span>
<span><span class="co"># will correctly specify cirular contigs for ppanggolin.  </span></span>
<span></span>
<span><span class="fu"><a href="reference/download_reference_genomes.html">download_reference_genomes</a></span><span class="op">(</span><span class="st">'LT2'</span>, <span class="st">'fna'</span>, <span class="st">'./reference_genomes/'</span><span class="op">)</span></span>
<span><span class="va">complete_genomes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">'./reference_genomes/'</span>, <span class="st">'.fna'</span>, full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">draft_genomes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">'./data/'</span>, <span class="st">'.fna'</span>, full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="reference/build_ppanggolin_file_fastas.html">build_ppanggolin_file_fastas</a></span><span class="op">(</span>complete_genome_paths <span class="op">=</span> <span class="va">complete_genomes</span>, </span>
<span>                             incomplete_genome_paths <span class="op">=</span> <span class="va">fna_files</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">write_tsv</span><span class="op">(</span><span class="st">'ppanggolin_file.tsv'</span>, col_names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="select-a-representative-set-of-isolates-from-a-pangenome">Select a representative set of isolates from a pangenome<a class="anchor" aria-label="anchor" href="#select-a-representative-set-of-isolates-from-a-pangenome"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in presence/absence matrix and format correctly:</span></span>
<span><span class="va">pan_PA</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">read_tsv</span><span class="op">(</span><span class="st">'./pan/gene_presence_absence.Rtab'</span><span class="op">)</span>  <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">column_to_rownames</span><span class="op">(</span>var <span class="op">=</span> <span class="st">'Gene'</span><span class="op">)</span>  <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># this will return a a small set of genomes that contain at least the proportion of</span></span>
<span><span class="co"># genes you specify</span></span>
<span></span>
<span><span class="co"># this set of genomes will contain 99% of all the genes detected in the pangenome</span></span>
<span><span class="fu"><a href="reference/get_pangenome_representatives.html">get_pangenome_representatives</a></span><span class="op">(</span>pan_mat <span class="op">=</span> <span class="va">pan_PA</span>, SEED <span class="op">=</span> <span class="fl">2</span>, desired_coverage <span class="op">=</span> <span class="fl">.99</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="todo">TODO<a class="anchor" aria-label="anchor" href="#todo"></a>
</h2>
<ul>
<li>Stats on available organisms, num clusters, most recent isolate <a href="https://www.ncbi.nlm.nih.gov/pathogens/organisms/" class="external-link uri">https://www.ncbi.nlm.nih.gov/pathogens/organisms/</a>
</li>
<li>update_collection() function?<ul>
<li>Should take and old metadata file and a new metadata file as inputs.<br>
</li>
<li>return a vector of new genomes that were not present in the old list.</li>
<li>also return genomes with newer assembly accession version<br>
</li>
</ul>
</li>
<li>extract_consensus_ag_species currently assumes isolates are from humans if the epi_type is clinical and no other information is available. This is probably wrong in some casesâ€¦</li>
<li>Download a more appropriate assembly summary file<ul><li>Or an assembly summary file for a species from the appropriate directory under genbank or refseq. e.g. <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/bacteria/Salmonella_enterica/assembly_summary.txt" class="uri">ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/bacteria/Salmonella_enterica/assembly_summary.txt</a>
</li></ul>
</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Jtrachsel/pdtools/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/Jtrachsel/pdtools/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3) + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing pdtools</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Julian Trachsel <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/https://orcid.org/0000-0003-2357-7737" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Jtrachsel/pdtools/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/Jtrachsel/pdtools/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://app.codecov.io/gh/Jtrachsel/pdtools?branch=main" class="external-link"><img src="https://codecov.io/gh/Jtrachsel/pdtools/branch/main/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Julian Trachsel.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
