<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ pdtools</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pdtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/tutorial_1.html">Downloading and organizing a collection of NCBI pathogen genomes</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="metadata-download-helpers">Metadata download helpers <a href="#metadata-download-helpers" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="list_PDGs.html">list_PDGs()</a></code> </p>
        </td>
        <td><p>List available PDG accessions for an organism</p></td>
      </tr><tr><td>
          <p><code><a href="list_organisms.html">list_organisms()</a></code> </p>
        </td>
        <td><p>List organisms available in the Pathogens database</p></td>
      </tr><tr><td>
          <p><code><a href="download_PDD_metadata.html">download_PDD_metadata()</a></code> </p>
        </td>
        <td><p>Download Pathogen Detection metadata for a given organism</p></td>
      </tr><tr><td>
          <p><code><a href="download_most_recent_complete.html">download_most_recent_complete()</a></code> </p>
        </td>
        <td><p>Download the most recent complete metadata for a specified organism</p></td>
      </tr><tr><td>
          <p><code><a href="download_gbk_assembly_summary.html">download_gbk_assembly_summary()</a></code> </p>
        </td>
        <td><p>Convenience function to download the assembly_summary.txt file from genbank</p></td>
      </tr><tr><td>
          <p><code><a href="find_most_recent_complete.html">find_most_recent_complete()</a></code> </p>
        </td>
        <td><p>Find most recent complete PDG</p></td>
      </tr><tr><td>
          <p><code><a href="check_complete_PDG.html">check_complete_PDG()</a></code> </p>
        </td>
        <td><p>Check if a single PDG is complete (AMR and Cluster metadata)</p></td>
      </tr><tr><td>
          <p><code><a href="make_ftp_paths.html">make_ftp_paths()</a></code> </p>
        </td>
        <td><p>generate ftp site paths for a selection of assembly accessions</p></td>
      </tr><tr><td>
          <p><code><a href="make_download_urls.html">make_download_urls()</a></code> </p>
        </td>
        <td><p>make specific ftp download paths for a dataframe with ftp_paths and assembly accessions</p></td>
      </tr><tr><td>
          <p><code><a href="make_SNPtree_urls.html">make_SNPtree_urls()</a></code> </p>
        </td>
        <td><p>generate ftp site download urls for all SNP trees containing the provided isolates</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="genome-download-helpers">Genome download helpers <a href="#genome-download-helpers" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="download_gbk_assembly_summary.html">download_gbk_assembly_summary()</a></code> </p>
        </td>
        <td><p>Convenience function to download the assembly_summary.txt file from genbank</p></td>
      </tr><tr><td>
          <p><code><a href="make_SNPtree_urls.html">make_SNPtree_urls()</a></code> </p>
        </td>
        <td><p>generate ftp site download urls for all SNP trees containing the provided isolates</p></td>
      </tr><tr><td>
          <p><code><a href="make_ftp_paths.html">make_ftp_paths()</a></code> </p>
        </td>
        <td><p>generate ftp site paths for a selection of assembly accessions</p></td>
      </tr><tr><td>
          <p><code><a href="make_download_urls.html">make_download_urls()</a></code> </p>
        </td>
        <td><p>make specific ftp download paths for a dataframe with ftp_paths and assembly accessions</p></td>
      </tr><tr><td>
          <p><code><a href="make_dest_paths.html">make_dest_paths()</a></code> </p>
        </td>
        <td><p>Make download destination paths</p></td>
      </tr><tr><td>
          <p><code><a href="download_genomes.html">download_genomes()</a></code> </p>
        </td>
        <td><p>Download specified files from NCBI ftp site</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="metadata-manipulations">Metadata manipulations <a href="#metadata-manipulations" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="extract_consensus_ag_species.html">extract_consensus_ag_species()</a></code> </p>
        </td>
        <td><p>Extract a consensus ag host species from metadata</p></td>
      </tr><tr><td>
          <p><code><a href="extract_earliest_year.html">extract_earliest_year()</a></code> </p>
        </td>
        <td><p>return a Year column containing the earliest year from the
available 'date' fields.</p></td>
      </tr><tr><td>
          <p><code><a href="extract_collection_agency.html">extract_collection_agency()</a></code> </p>
        </td>
        <td><p>extract collecting agency from several metadata fields</p></td>
      </tr><tr><td>
          <p><code><a href="extract_country.html">extract_country()</a></code> </p>
        </td>
        <td><p>Extract a standardized country name from geo_loc_tag column</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="pangenome-tools">Pangenome tools <a href="#pangenome-tools" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="build_ppanggolin_file_fastas.html">build_ppanggolin_file_fastas()</a></code> </p>
        </td>
        <td><p>Build a ppanggolin file from fastas</p></td>
      </tr><tr><td>
          <p><code><a href="generate_genome_vector.html">generate_genome_vector()</a></code> </p>
        </td>
        <td><p>Generate a genome tibble containing presence/absence of genes in a fake pangenome</p></td>
      </tr><tr><td>
          <p><code><a href="generate_pangenome.html">generate_pangenome()</a></code> </p>
        </td>
        <td><p>Generate a random synthetic pangenome gene presence absence matrix</p></td>
      </tr><tr><td>
          <p><code><a href="pan_mat_to_gene_vec_tibble.html">pan_mat_to_gene_vec_tibble()</a></code> </p>
        </td>
        <td><p>Convert a pangenome PA matrix to a tibble of gene vectors</p></td>
      </tr><tr><td>
          <p><code><a href="get_pangenome_representatives.html">get_pangenome_representatives()</a></code> </p>
        </td>
        <td><p>Select a minimum set of genomes that best represent the gene content of the pangenome</p></td>
      </tr><tr><td>
          <p><code><a href="get_pangenome_representatives2.html">get_pangenome_representatives2()</a></code> </p>
        </td>
        <td><p>Get pangenome representatives from a gene_vec_tibble</p></td>
      </tr><tr><td>
          <p><code><a href="remove_strict_core.html">remove_strict_core()</a></code> </p>
        </td>
        <td><p>Removes genes present in all genomes from pangenome presence/absence matrix</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="data">Data <a href="#data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="klebsiella_example_dat.html">klebsiella_example_dat</a></code> </p>
        </td>
        <td><p>NCBI Pathogen Detection metadata for 200 Klebsiella isolates</p></td>
      </tr><tr><td>
          <p><code><a href="country_vector.html">country_vector</a></code> </p>
        </td>
        <td><p>A named vector of country names</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Julian Trachsel.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

