<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ pdtools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">pdtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Jtrachsel/pdtools/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="metadata-access">Metadata Access<a class="anchor" aria-label="anchor" href="#metadata-access"></a></h2>
      
      <p class="section-desc"></p><p>Functions for downloading various types of metadata from the NCBI Pathogens database</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="list_organisms.html">list_organisms()</a></code> 
        </dt>
        <dd>List organisms available in the Pathogens database</dd>
      </dl><dl><dt>
          
          <code><a href="list_PDGs.html">list_PDGs()</a></code> 
        </dt>
        <dd>List available PDG accessions for an organism</dd>
      </dl><dl><dt>
          
          <code><a href="download_most_recent_complete.html">download_most_recent_complete()</a></code> 
        </dt>
        <dd>Download the most recent complete metadata for a specified organism</dd>
      </dl><dl><dt>
          
          <code><a href="download_PDD_metadata.html">download_PDD_metadata()</a></code> 
        </dt>
        <dd>Download Pathogen Detection metadata for a given organism</dd>
      </dl><dl><dt>
          
          <code><a href="get_PDG_version.html">get_PDG_version()</a></code> 
        </dt>
        <dd>get the version of the PDD metadata</dd>
      </dl><dl><dt>
          
          <code><a href="download_gbk_assembly_summary.html">download_gbk_assembly_summary()</a></code> 
        </dt>
        <dd>Convenience function to download the assembly_summary.txt file from genbank</dd>
      </dl></div><div class="section level2">
      <h2 id="genome-downloads">Genome downloads<a class="anchor" aria-label="anchor" href="#genome-downloads"></a></h2>
      
      <p class="section-desc"></p><p>Functions for organizing and downloading pathogen genomes.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="make_ftp_paths.html">make_ftp_paths()</a></code> 
        </dt>
        <dd>generate ftp site paths for a selection of assembly accessions</dd>
      </dl><dl><dt>
          
          <code><a href="make_download_urls.html">make_download_urls()</a></code> 
        </dt>
        <dd>make specific ftp download paths for a dataframe with ftp_paths and assembly accessions</dd>
      </dl><dl><dt>
          
          <code><a href="make_dest_paths.html">make_dest_paths()</a></code> 
        </dt>
        <dd>Make download destination paths</dd>
      </dl><dl><dt>
          
          <code><a href="download_genomes.html">download_genomes()</a></code> 
        </dt>
        <dd>Download a set of specified genome files</dd>
      </dl><dl><dt>
          
          <code><a href="download_reference_genomes.html">download_reference_genomes()</a></code> 
        </dt>
        <dd>Download reference genomes</dd>
      </dl></div><div class="section level2">
      <h2 id="metadata-cleanup-tools">Metadata cleanup tools<a class="anchor" aria-label="anchor" href="#metadata-cleanup-tools"></a></h2>
      
      <p class="section-desc"></p><p>Functions for extracting information from provided metadata</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="extract_collection_agency.html">extract_collection_agency()</a></code> 
        </dt>
        <dd>extract collecting agency from several metadata fields</dd>
      </dl><dl><dt>
          
          <code><a href="extract_consensus_ag_species.html">extract_consensus_ag_species()</a></code> 
        </dt>
        <dd>Extract a consensus ag host species from metadata</dd>
      </dl><dl><dt>
          
          <code><a href="extract_country.html">extract_country()</a></code> 
        </dt>
        <dd>Extract a standardized country name from geo_loc_tag column</dd>
      </dl><dl><dt>
          
          <code><a href="extract_state.html">extract_state()</a></code> 
        </dt>
        <dd>Extract the state name from the geo_loc_name column, this only makes sence for USA isolates</dd>
      </dl><dl><dt>
          
          <code><a href="extract_earliest_year.html">extract_earliest_year()</a></code> 
        </dt>
        <dd>return a Year column containing the earliest year from the
available 'date' fields.</dd>
      </dl></div><div class="section level2">
      <h2 id="pangenome-tools">Pangenome tools<a class="anchor" aria-label="anchor" href="#pangenome-tools"></a></h2>
      
      <p class="section-desc"></p><p>Functions for working with pangenomes.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="build_ppanggolin_file_fastas.html">build_ppanggolin_file_fastas()</a></code> 
        </dt>
        <dd>Build a ppanggolin file from fastas</dd>
      </dl><dl><dt>
          
          <code><a href="cluster_genomes.html">cluster_genomes()</a></code> 
        </dt>
        <dd>Cluster genomes at 4 levels from a pangenome gene PA matrix
Needs parallelDist, igraph,
genomes as rows and genes as columns?</dd>
      </dl><dl><dt>
          
          <code><a href="mark_outliers.html">mark_outliers()</a></code> 
        </dt>
        <dd>Mark outliers from a distance matrix</dd>
      </dl><dl><dt>
          
          <code><a href="remove_strict_core.html">remove_strict_core()</a></code> 
        </dt>
        <dd>Removes genes present in all genomes from pangenome presence/absence matrix</dd>
      </dl><dl><dt>
          
          <code><a href="calculate_novelty.html">calculate_novelty()</a></code> 
        </dt>
        <dd>Calculate genome 'novelty' from a list of selection sets returned by pick_derep_sets</dd>
      </dl><dl><dt>
          
          <code><a href="get_pangenome_representatives.html">get_pangenome_representatives()</a></code> 
        </dt>
        <dd>Select a minimum set of genomes that best represent the gene content of the pangenome</dd>
      </dl><dl><dt>
          
          <code><a href="get_pangenome_representatives_jaccard.html">get_pangenome_representatives_jaccard()</a></code> 
        </dt>
        <dd>Get a subset of a pangenome that tries to represent the gene presence absence diversity
Selects a random genome,
then selects the most distant genome from the selected genome
then</dd>
      </dl><dl><dt>
          
          <code><a href="pick_derep_sets.html">pick_derep_sets()</a></code> 
        </dt>
        <dd>Pick multiple de-replication sets from a pangenome, uses furrr::future_map
Be sure to set your 'plan()'!!!</dd>
      </dl></div><div class="section level2">
      <h2 id="snp-tree-downloads">SNP tree downloads<a class="anchor" aria-label="anchor" href="#snp-tree-downloads"></a></h2>
      
      <p class="section-desc"></p><p>functions to organize and download SNP trees from NCBI</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="make_SNPtree_urls.html">make_SNPtree_urls()</a></code> 
        </dt>
        <dd>generate ftp site download urls for all SNP trees containing the provided isolates</dd>
      </dl><dl><dt>
          
          <code><a href="make_SNP_tree_dest.html">make_SNP_tree_dest()</a></code> 
        </dt>
        <dd>Make SNP tree download destination paths</dd>
      </dl><dl><dt>
          
          <code><a href="download_SNP_trees.html">download_SNP_trees()</a></code> 
        </dt>
        <dd>Download SNP trees from a dataframe containing SNP_tree_urls and dests</dd>
      </dl></div><div class="section level2">
      <h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a></h2>
      
      <p class="section-desc"></p><p>included data</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="country_vector.html">country_vector</a></code> 
        </dt>
        <dd>A named vector of country names</dd>
      </dl><dl><dt>
          
          <code><a href="klebsiella_example_dat.html">klebsiella_example_dat</a></code> 
        </dt>
        <dd>NCBI Pathogen Detection metadata for 200 Klebsiella isolates</dd>
      </dl><dl><dt>
          
          <code><a href="example_pangenome_matrix.html">example_pangenome_matrix</a></code> 
        </dt>
        <dd>A gene presence absence matrix for a synthetic pangenome</dd>
      </dl><dl><dt>
          
          <code><a href="example_pan_dist.html">example_pan_dist</a></code> 
        </dt>
        <dd>A distance matrix describing the jaccard distance between genomes in the
provided example pangenome</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Julian Trachsel.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

